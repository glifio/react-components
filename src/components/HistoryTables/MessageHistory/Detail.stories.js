import { ApolloProvider } from '@apollo/client'
import { client } from '../apolloClient'
import theme from '../../theme'
import ThemeProvider from '../../ThemeProvider'

import MessageDetail from './Detail'

export default {
  title: 'MessageHistory/MessageDetail',
  component: MessageDetail,
  decorators: [
    Story => (
      <ApolloProvider client={client}>
        <div style={{ display: 'flex', justifyContent: 'center' }}>
          <ThemeProvider theme={theme}>{Story()}</ThemeProvider>
        </div>
      </ApolloProvider>
    )
  ],
  parameters: { actions: { argTypesRegex: '^on.*' } }
}

const Template = args => <MessageDetail {...args} />

export const Detail = Template.bind({})
Detail.args = {
  cid: 'bafy2bzacebf6orovs3hpx6teptxanwighhq7aui4b5vqwbgkq6vl5qc324cv4'
}
