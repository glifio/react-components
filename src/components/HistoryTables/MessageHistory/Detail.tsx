import React, { useState, useMemo, useEffect } from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import Link from 'next/link'
import * as dayjs from 'dayjs'
import * as relativeTime from 'dayjs/plugin/relativeTime'
import {
  useMessageQuery,
  useChainHeadSubscription
} from '../../../generated/graphql'
import Box from '../../Box'
import { IconClock } from '../../Icons'
import { P, HR } from '../../Typography'
import { Badge } from '../generic'
import { Head, Line, Status, Confirmations, Parameters } from '../detail'
import {
  attoFilToFil,
  getTotalCost,
  getGasPercentage,
  formatNumber
} from '../utils'

// add RelativeTime plugin to Day.js
dayjs.extend(relativeTime.default)

const SeeMore = styled(P).attrs(() => ({
  color: 'core.primary',
  role: 'button'
}))`
  cursor: pointer;
`

export default function MessageDetail(props: MessageDetailProps) {
  const { cid, speedUp, cancel, addressHref, confirmations } = props
  const [time, setTime] = useState(Date.now())
  const [seeMore, setSeeMore] = useState(false)
  const { data, loading, error } = useMessageQuery({
    variables: { cid }
  })
  const chainHeadSubscription = useChainHeadSubscription({
    variables: {}
  })
  const value = useMemo(
    () => (data?.message.value ? attoFilToFil(data?.message.value) : ''),
    [data?.message.value]
  )
  const totalCost = useMemo(
    () => (data?.message ? getTotalCost(data.message) : ''),
    [data?.message]
  )
  const gasPercentage = useMemo(
    () => (data?.message ? getGasPercentage(data.message) : ''),
    [data?.message]
  )
  const timestamp = useMemo(
    () => data?.message.block.Timestamp ?? null,
    [data?.message.block.Timestamp]
  )
  const date = useMemo(
    () => (timestamp ? dayjs.unix(timestamp).toString() : ''),
    [timestamp]
  )
  const age = useMemo(
    () => (timestamp ? dayjs.unix(timestamp).from(time) : ''),
    [timestamp, time]
  )
  const confirmationCount = useMemo(
    () =>
      chainHeadSubscription.data?.chainHead.height && data?.message.height
        ? chainHeadSubscription.data.chainHead.height - data.message.height
        : 0,
    [data?.message.height, chainHeadSubscription.data?.chainHead.height]
  )

  useEffect(() => {
    const interval = setInterval(() => setTime(Date.now()), 1000)
    return () => clearInterval(interval)
  })

  if (loading) return <p>Loading...</p>
  if (error) return <p>Error :( {error.message}</p>

  return (
    <Box>
      <Head title='Message Overview' speedUp={speedUp} cancel={cancel} />
      <HR />
      <Line label='CID'>{cid}</Line>
      <Line label='Status and Confirmations'>
        <Status exitCode={data.message.exitCode} />
        <Confirmations count={confirmationCount} total={confirmations} />
      </Line>
      <Line label='Height'>{data.message.height}</Line>
      <Line label='Timestamp'>
        <IconClock width='1.125em' />
        {age} ({date})
      </Line>
      <HR />
      <Line label='From'>
        {data.message.from.robust}
        <Link
          href={addressHref(data.message.from.robust)}
        >{`(${data.message.from.id})`}</Link>
      </Line>
      <Line label='To'>
        {data.message.to.robust}
        <Link
          href={addressHref(data.message.to.robust)}
        >{`(${data.message.to.id})`}</Link>
      </Line>
      <HR />
      <Line label='Value'>{value}</Line>
      <Line label='Transaction Fee'>{totalCost}</Line>
      <Line label='Method'>
        <Badge color='purple'>{data.message.methodName.toUpperCase()}</Badge>
      </Line>
      <HR />
      <SeeMore onClick={() => setSeeMore(!seeMore)}>
        Click to see {seeMore ? 'less ↑' : 'more ↓'}
      </SeeMore>
      <HR />
      {seeMore && (
        <>
          <Line label='Gas Limit & Usage by Txn'>
            {formatNumber(data.message.gasLimit)}
            <span className='gray'>|</span>
            {formatNumber(data.message.gasUsed)} attoFil
            <span>({gasPercentage})</span>
          </Line>
          <Line label='Gas Fees'>
            <span className='gray'>Premium</span>
            {formatNumber(data.message.gasPremium)} attoFIL
          </Line>
          <Line label=''>
            <span className='gray'>Fee Cap</span>
            {formatNumber(data.message.gasFeeCap)} attoFIL
          </Line>
          <Line label=''>
            <span className='gray'>Base</span>
            {formatNumber(data.message.baseFeeBurn)} attoFIL
          </Line>
          <Line label='Gas Burnt'>
            {formatNumber(data.message.gasBurned)} attoFIL
          </Line>
          <HR />
          <Parameters params={{ params: data.message.params }} depth={0} />
        </>
      )}
    </Box>
  )
}

type MessageDetailProps = {
  cid: string
  speedUp?: () => void
  cancel?: () => void
  addressHref: (address: string) => string
  confirmations: number
}

MessageDetail.propTypes = {
  cid: PropTypes.string.isRequired,
  speedUp: PropTypes.func,
  cancel: PropTypes.func,
  addressHref: PropTypes.func.isRequired,
  confirmations: PropTypes.number.isRequired
}
