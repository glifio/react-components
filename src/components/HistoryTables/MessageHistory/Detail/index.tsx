import { useState, useMemo, useEffect } from 'react'
import { ExecutionTrace } from '@glif/filecoin-wallet-provider'
import { FilecoinNumber } from '@glif/filecoin-number'
import PropTypes from 'prop-types'
import { Message, useStateReplayQuery } from '../../../../generated/graphql'
import { AddressLink } from '../../../LabeledText/AddressLink'
import { DetailCaption, MessageDetailBase, SeeMoreContent } from '../../detail'
import { useMessage } from '../hooks/useAllMessages'
import { useMethodName } from '../hooks/useMethodName'
import { Lines, Line, StandardBox, PageTitle } from '../../../Layout'
import { isAddrEqual, makeFriendlyBalance } from '../../../../utils'
import {
  ExecReturn,
  getAddrFromReceipt
} from '../../../../utils/getAddrFromReceipt'
import { LoadingIcon } from '../../../Loading/LoadingIcon'
import { ButtonV2Link } from '../../../Button/V2'
import { IconCancel, IconSpeedUp } from '../../../Icons'
import {
  useEnvironment,
  useLogger
} from '../../../../services/EnvironmentProvider'

enum MessageState {
  Loading,
  // message is in the mempool
  Pending,
  // message is included in chain, but not yet executed (no gas cost, no receipt)
  Confirmed,
  // message has been fully executed - all variables should exist now
  Executed,
  // failed to get message
  Error,
  NotFound
}

export default function MessageDetail(props: MessageDetailProps) {
  const { cid, speedUpHref, cancelHref, confirmations } = props
  const time = useMemo(() => Date.now(), [])
  const [seeMore, setSeeMore] = useState(false)
  const { message, error, loading, pending } = useMessage(cid)
  const { isProd } = useEnvironment()
  const logger = useLogger()

  const { data: stateReplayQuery, error: stateReplayError } =
    useStateReplayQuery({
      variables: { cid: props.cid },
      pollInterval: 30000,
      fetchPolicy: 'cache-first',
      // give the message time to execute on-chain before fetching
      skip: !props.cid || props.confirmations < 2
    })

  const transactionFee = useMemo<string>(() => {
    if (pending) return 'Pending...'
    const cost = stateReplayQuery?.stateReplay?.gasCost?.totalCost
    return cost
      ? `${makeFriendlyBalance(new FilecoinNumber(cost, 'attofil'))} FIL`
      : 'Calculating...'
  }, [stateReplayQuery?.stateReplay?.gasCost?.totalCost, pending])

  const gasUsed = useMemo(
    () => stateReplayQuery?.stateReplay?.gasCost?.gasUsed || 0,
    [stateReplayQuery?.stateReplay?.gasCost]
  )

  const executionTrace = useMemo(
    () =>
      (stateReplayQuery?.stateReplay?.executionTrace
        ?.executionTrace as unknown as ExecutionTrace) || null,
    [stateReplayQuery]
  )

  const { methodName, actorName } = useMethodName(message)

  const { coinType } = useEnvironment()

  const execReturn = useMemo<ExecReturn | null>(() => {
    if (!message) return null
    const isToExec = isAddrEqual(message?.to, 'f01')
    const isInitTx = Number(message?.method) === 2
    const receiptReturn = stateReplayQuery?.stateReplay?.receipt?.return
    return isToExec && isInitTx && receiptReturn
      ? getAddrFromReceipt(receiptReturn, coinType)
      : null
  }, [message, stateReplayQuery?.stateReplay?.receipt, coinType])

  const messageState = useMemo<MessageState>(() => {
    if (error) return MessageState.Error
    if (loading) return MessageState.Loading
    if (pending) return MessageState.Pending
    if (!message) return MessageState.NotFound
    if (!gasUsed) return MessageState.Confirmed
    return MessageState.Executed
  }, [pending, message, error, gasUsed, loading])

  // Log errors
  useEffect(() => error && logger.error(error), [error, logger])
  useEffect(
    () => stateReplayError && logger.error(stateReplayError),
    [stateReplayError, logger]
  )

  return (
    <>
      <PageTitle
        sideContent={
          pending && (
            <>
              {speedUpHref && !isProd && (
                <ButtonV2Link green href={speedUpHref} retainParams>
                  <IconSpeedUp width='1.25rem' />
                  Speed up
                </ButtonV2Link>
              )}
              {cancelHref && !isProd && (
                <ButtonV2Link red href={cancelHref} retainParams>
                  <IconCancel width='0.8rem' />
                  Cancel
                </ButtonV2Link>
              )}
            </>
          )
        }
      >
        Message Overview
      </PageTitle>
      <hr />
      <Lines>
        {messageState === MessageState.Error && (
          <DetailCaption
            name='Message Overview'
            caption='Error'
            loading={false}
            error={error}
          />
        )}
        {messageState === MessageState.NotFound && (
          <StandardBox>
            <p>Message {cid} not found.</p>
            <p>
              Note - it may take 1-2 minutes for a recently confirmed message to
              show up here.
            </p>
          </StandardBox>
        )}
        {messageState === MessageState.Loading && (
          <StandardBox>
            <LoadingIcon />
            <p>Searching for {cid}... Give us a moment</p>
          </StandardBox>
        )}
        {messageState === MessageState.Pending && (
          <MessageDetailBase
            cid={cid}
            methodName={methodName}
            message={message}
            time={time}
            pending
          />
        )}
        {messageState === MessageState.Confirmed && (
          <>
            <MessageDetailBase
              cid={cid}
              methodName={methodName}
              confirmations={confirmations}
              time={time}
              message={message}
            />
            <p>Loading more details...</p>
          </>
        )}
        {messageState === MessageState.Executed && (
          <>
            <MessageDetailBase
              cid={cid}
              exitCode={stateReplayQuery?.stateReplay?.receipt?.exitCode}
              methodName={methodName}
              confirmations={confirmations}
              time={time}
              message={message}
            />
            <Line label='Transaction Fee'>{transactionFee}</Line>
            {!!execReturn && (
              <Line label='New actor created: '>
                <AddressLink
                  id={execReturn.id}
                  address={execReturn.robust}
                  hideCopyText={false}
                />
              </Line>
            )}
            <hr />
            {seeMore ? (
              <p role='button' onClick={() => setSeeMore(false)}>
                Click to see less ↑
              </p>
            ) : (
              <p role='button' onClick={() => setSeeMore(true)}>
                Click to see more ↓
              </p>
            )}
            {seeMore && (
              <SeeMoreContent
                message={message as Message}
                gasUsed={gasUsed}
                gasCost={stateReplayQuery?.stateReplay?.gasCost}
                executionTrace={executionTrace}
                actorName={actorName}
              />
            )}
          </>
        )}
      </Lines>
    </>
  )
}

type MessageDetailProps = {
  cid: string
  speedUpHref?: string
  cancelHref?: string
  confirmations: number
}

MessageDetail.propTypes = {
  cid: PropTypes.string.isRequired,
  speedUpHref: PropTypes.string,
  cancelHref: PropTypes.string,
  confirmations: PropTypes.number
}

MessageDetail.defaultProps = {
  confirmations: 50
}
