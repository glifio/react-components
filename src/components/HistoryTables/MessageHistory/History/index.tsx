import { useMemo } from 'react'
import PropTypes from 'prop-types'
import Box from '../../../Box'
import MessageConfirmedRow from './MessageConfirmedRow'
import MessagePendingRow from './MessagePendingRow'
import { MessageRowColumnTitles } from './MessageRowColumnTitles'
import { ADDRESS_PROPTYPE } from '../../../../customPropTypes'
import { ButtonV2 } from '../../../Button/V2'
import { useAllMessages } from '../hooks/useAllMessages'
import { Caption, PageTitle } from '../../../Layout'
import { SyncStatus } from '../../../SyncStatus'

export default function MessageHistoryTable(props: MessageHistoryTableProps) {
  const {
    messages,
    pendingMsgs,
    loading,
    error,
    fetchMore,
    fetchingMore,
    lastPage
  } = useAllMessages(props.address, props.offset)

  const isEmpty = useMemo(
    () => !pendingMsgs?.length && !messages?.length,
    [pendingMsgs, messages]
  )

  return (
    <Box>
      <PageTitle>Transaction History</PageTitle>
      {props.warnMissingData && <SyncStatus />}
      <br />
      <table>
        <Caption
          name='Transaction History'
          loading={isEmpty && loading}
          error={error}
          empty={isEmpty}
        />
        <MessageRowColumnTitles />
        <tbody>
          {pendingMsgs?.map(message => (
            <MessagePendingRow
              key={message.cid}
              message={message}
              cidHref={props.cidHref}
              inspectingAddress={props.address}
            />
          ))}
          {messages?.map(message => (
            <MessageConfirmedRow
              key={message.cid}
              message={message}
              cidHref={props.cidHref}
              inspectingAddress={props.address}
            />
          ))}
        </tbody>
      </table>
      {!isEmpty && !lastPage && (
        <Box pt='4.5rem' textAlign='center'>
          <ButtonV2
            onClick={fetchMore}
            px='18rem'
            disabled={loading || fetchingMore}
          >
            {loading || fetchingMore ? 'Loading...' : 'Load more'}
          </ButtonV2>
        </Box>
      )}
    </Box>
  )
}

type MessageHistoryTableProps = {
  offset: number
  address: string
  warnMissingData?: boolean
  cidHref: (cid: string) => string
}

MessageHistoryTable.propTypes = {
  address: ADDRESS_PROPTYPE.isRequired,
  cidHref: PropTypes.func.isRequired,
  offset: PropTypes.number,
  warnMissingData: PropTypes.bool
}

MessageHistoryTable.defaultProps = {
  offset: 0
}
