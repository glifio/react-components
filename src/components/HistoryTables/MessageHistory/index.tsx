import { useMemo } from 'react'
import PropTypes from 'prop-types'
import Box from '../../Box'
import MessageConfirmedRow from './MessageConfirmedRow'
import MessagePendingRow from './MessagePendingRow'
import { MessageRowColumnTitles } from './MessageRowColumnTitles'
import { ADDRESS_PROPTYPE } from '../../../customPropTypes'
import ButtonV2 from '../../Button/V2'
import { TABLE, TableCaption } from '../table'
import { useAllMessages } from '../useAllMessages'
import { Title } from '../generic'

const DEFAULT_LIMIT = 10

export default function MessageHistoryTable(props: MessageHistoryTableProps) {
  const {
    messages,
    pendingMsgs,
    loading,
    error,
    chainHeadSub,
    fetchMore,
    fetchingMore
  } = useAllMessages(props.address, props.offset)

  const lastPage = useMemo(
    () => messages?.length < DEFAULT_LIMIT,
    [messages?.length]
  )

  const isEmpty = useMemo(
    () => !pendingMsgs?.length && !messages?.length,
    [pendingMsgs, messages]
  )

  return (
    <Box>
      <Title>Transaction History</Title>
      <TABLE className='narrow'>
        <TableCaption
          name='Transaction History'
          loading={isEmpty && loading}
          error={error}
          empty={isEmpty}
        />
        <MessageRowColumnTitles />
        <tbody>
          {pendingMsgs?.map(message => (
            <MessagePendingRow
              key={message.cid}
              message={message}
              cidHref={props.cidHref}
              inspectingAddress={props.address}
            />
          ))}
          {messages?.map(message => (
            <MessageConfirmedRow
              key={message.cid}
              message={message}
              cidHref={props.cidHref}
              inspectingAddress={props.address}
              chainHeadSub={chainHeadSub}
            />
          ))}
        </tbody>
      </TABLE>
      {!isEmpty && !lastPage && (
        <Box pt='4.5rem' textAlign='center'>
          <ButtonV2 onClick={fetchMore} px='18rem' disabled={loading || fetchingMore}>
            {loading || fetchingMore ? 'Loading...' : 'Load more'}
          </ButtonV2>
        </Box>
      )}
    </Box>
  )
}

type MessageHistoryTableProps = {
  offset: number
  address: string
  cidHref: (cid: string, height?: string) => string
}

MessageHistoryTable.propTypes = {
  offset: PropTypes.number,
  address: ADDRESS_PROPTYPE,
  cidHref: PropTypes.func
}

MessageHistoryTable.defaultProps = {
  offset: 0,
  // TODO
  cidHref: (cid: string) => `/#/detail/${cid}`
}
