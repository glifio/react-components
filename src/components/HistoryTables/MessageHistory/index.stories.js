import { ApolloProvider } from '@apollo/client'
import { client } from '../apolloClient'
import theme from '../../theme'
import ThemeProvider from '../../ThemeProvider'

import MessageHistory from './index'

export default {
  title: 'MessageHistory/Table',
  component: MessageHistory,
  decorators: [
    Story => (
      <ApolloProvider client={client}>
        <ThemeProvider theme={theme}>{Story()}</ThemeProvider>
      </ApolloProvider>
    )
  ],
  parameters: { actions: { argTypesRegex: '^on.*' } }
}

const Template = args => (
  <MessageHistory cidHref={cid => `/#/detail/${cid}`} {...args} />
)

export const Base = Template.bind({})
Base.args = {
  address: 't1iuryu3ke2hewrcxp4ezhmr5cmfeq3wjhpxaucza'
}
