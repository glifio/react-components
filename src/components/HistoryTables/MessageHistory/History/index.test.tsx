import React from 'react'
import { render, act, screen } from '@testing-library/react'
import { MockedProvider } from '@apollo/client/testing'
import ThemeProvider from '../../../ThemeProvider'
import theme from '../../../theme'
import MessageHistory from '.'
import {
  WALLET_ADDRESS,
  WALLET_ADDRESS_2,
  WALLET_ID,
  WALLET_ID_2
} from '../../../../test-utils/constants'
import { MessageConfirmedRow } from '../../types'
import { TestEnvironment } from '../../../../test-utils/TestEnvironment'

jest.mock('dayjs')

describe('Message history', () => {
  test('it renders no message history correctly with historical data missing warning', async () => {
    jest
      .spyOn(require('../hooks/useAllMessages'), 'useAllMessages')
      .mockImplementation(() => {
        return {
          error: undefined,
          loading: false,
          fetchMore: () => {},
          fetchingMore: false,
          pendingMsgs: [],
          messages: [],
          lastPage: true
        }
      })

    let container

    act(() => {
      const res = render(
        <MockedProvider>
          <ThemeProvider theme={theme}>
            <MessageHistory
              offset={0}
              address={WALLET_ADDRESS}
              cidHref={cid => `/message/?cid=${cid}`}
              warnMissingData
            />
          </ThemeProvider>
        </MockedProvider>
      )
      container = res.container
    })

    expect(screen.getByText(/No records found/))
    expect(screen.getByText(/Syncing data from Mainnet/))
    expect(container).toMatchSnapshot()
  })

  test('it renders messages', async () => {
    const makeMessages = (count: number): MessageConfirmedRow[] => {
      const messages = [] as MessageConfirmedRow[]
      for (let i = 0; i < count; i++) {
        messages.push({
          cid: `bafy2bzaced3ub5g4v35tj7n74zsog3dmcum4tk${i}qmchbhjx7q747jghal3l4g`,
          from: {
            id: WALLET_ID,
            robust: WALLET_ADDRESS
          },
          to: {
            id: WALLET_ID_2,
            robust: WALLET_ADDRESS_2
          },
          height: 1020,
          value: '1000',
          method: '0'
        })
      }
      return messages
    }
    jest
      .spyOn(require('../hooks/useAllMessages'), 'useAllMessages')
      .mockImplementation(() => {
        return {
          error: undefined,
          loading: false,
          fetchMore: () => {},
          fetchingMore: false,
          pendingMsgs: [],
          messages: makeMessages(5),
          lastPage: true
        }
      })

    let container

    act(() => {
      const res = render(
        <TestEnvironment>
          <MockedProvider>
            <ThemeProvider theme={theme}>
              <MessageHistory
                offset={0}
                address={WALLET_ADDRESS}
                cidHref={cid => `/message/?cid=${cid}`}
                warnMissingData
              />
            </ThemeProvider>
          </MockedProvider>
        </TestEnvironment>
      )
      container = res.container
    })

    expect(screen.getByText(/Transaction History/))
    expect(screen.getByText(/Syncing data from Mainnet/))
    expect(container).toMatchSnapshot()
  })
})
