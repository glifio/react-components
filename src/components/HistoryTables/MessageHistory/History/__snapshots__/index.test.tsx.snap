// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message history it renders messages 1`] = `
.c5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: var(--space-m);
  padding: 0 1em;
  height: 2em;
  line-height: 2em;
  border-radius: 1em;
  text-align: center;
  white-space: nowrap;
  color: var(--yellow-dark);
  background-color: var(--yellow-light);
}

.c3 h4 {
  margin: 0;
  color: var(--gray-dark);
}

.c3 p {
  margin: 0;
  line-height: 1.5;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  grid-gap: 0.25em;
  line-height: 1.5;
}

.c4 > a {
  color: var(--purple-medium);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4 > a:hover {
  color: var(--purple-medium);
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c4 > a svg {
  -webkit-transition: 0.24s ease-in-out;
  transition: 0.24s ease-in-out;
  vertical-align: middle;
}

.c4 > a svg:hover {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
  fill: var(--purple-medium);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  gap: var(--space-l);
  margin: 1em 0;
  font-size: var(--font-size-xl);
}

.c0 .side-bar {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: var(--space-m);
  font-size: 1rem;
}

.c1 {
  margin: 0;
  color: var(--purple-medium);
}

.c2 {
  color: var(--yellow-dark);
}

<div>
  <div>
    <div
      class="c0"
    >
      <h2
        class="c1"
      >
        Transaction History
      </h2>
      <div
        class="side-bar"
      >
        <svg
          fill="none"
          height="20px"
          viewBox="0 0 22 20"
          width="22px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 7.74982V9.99986M11 14.4999H11.0112M3.24282 19H18.7572C20.4815 19 21.5586 17.1246 20.6965 15.6249L12.9393 2.12474C12.0771 0.625088 9.92286 0.625088 9.0607 2.12474L1.30352 15.6249C0.441363 17.1246 1.5185 19 3.24282 19Z"
            stroke="var(--yellow-dark)"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <span
          class="c2"
        >
          Syncing data from Mainnet
        </span>
      </div>
    </div>
    <table>
      <thead>
        <tr>
          <th>
            CID
          </th>
          <th>
            Method
          </th>
          <th>
            Height
          </th>
          <th>
            Age
          </th>
          <th>
            From
          </th>
          <th />
          <th>
            To
          </th>
          <th>
            Value
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <a
              href="/message?cid=bafy2bzaced3ub5g4v35tj7n74zsog3dmcum4tk0qmchbhjx7q747jghal3l4g"
            >
              bafy2b ... al3l4g
            </a>
          </td>
          <td>
            send
          </td>
          <td>
            1020
          </td>
          <td>
            2022-08-02 09:50:30
          </td>
          <td>
            <div
              class="c3"
            >
              <div
                class="c4"
                color="var(--purple-medium)"
              >
                <span>
                  t1z225 ... 6z6wgi
                </span>
              </div>
            </div>
          </td>
          <td>
            <div
              class="c5"
              color="yellow"
            >
              <span>
                OUT
              </span>
            </div>
          </td>
          <td>
            <div
              class="c3"
            >
              <div
                class="c4"
                color="var(--purple-medium)"
              >
                <a
                  href="https://explorer-calibration.glif.link/actor/?address=t1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a"
                  rel="noreferrer noopener"
                  target="_blank"
                >
                  t1nq5k ... btuo3a
                </a>
              </div>
            </div>
          </td>
          <td>
            0.000000000000001 FIL
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/message?cid=bafy2bzaced3ub5g4v35tj7n74zsog3dmcum4tk1qmchbhjx7q747jghal3l4g"
            >
              bafy2b ... al3l4g
            </a>
          </td>
          <td>
            send
          </td>
          <td>
            1020
          </td>
          <td>
            2022-08-02 09:50:30
          </td>
          <td>
            <div
              class="c3"
            >
              <div
                class="c4"
                color="var(--purple-medium)"
              >
                <span>
                  t1z225 ... 6z6wgi
                </span>
              </div>
            </div>
          </td>
          <td>
            <div
              class="c5"
              color="yellow"
            >
              <span>
                OUT
              </span>
            </div>
          </td>
          <td>
            <div
              class="c3"
            >
              <div
                class="c4"
                color="var(--purple-medium)"
              >
                <a
                  href="https://explorer-calibration.glif.link/actor/?address=t1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a"
                  rel="noreferrer noopener"
                  target="_blank"
                >
                  t1nq5k ... btuo3a
                </a>
              </div>
            </div>
          </td>
          <td>
            0.000000000000001 FIL
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/message?cid=bafy2bzaced3ub5g4v35tj7n74zsog3dmcum4tk2qmchbhjx7q747jghal3l4g"
            >
              bafy2b ... al3l4g
            </a>
          </td>
          <td>
            send
          </td>
          <td>
            1020
          </td>
          <td>
            2022-08-02 09:50:30
          </td>
          <td>
            <div
              class="c3"
            >
              <div
                class="c4"
                color="var(--purple-medium)"
              >
                <span>
                  t1z225 ... 6z6wgi
                </span>
              </div>
            </div>
          </td>
          <td>
            <div
              class="c5"
              color="yellow"
            >
              <span>
                OUT
              </span>
            </div>
          </td>
          <td>
            <div
              class="c3"
            >
              <div
                class="c4"
                color="var(--purple-medium)"
              >
                <a
                  href="https://explorer-calibration.glif.link/actor/?address=t1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a"
                  rel="noreferrer noopener"
                  target="_blank"
                >
                  t1nq5k ... btuo3a
                </a>
              </div>
            </div>
          </td>
          <td>
            0.000000000000001 FIL
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/message?cid=bafy2bzaced3ub5g4v35tj7n74zsog3dmcum4tk3qmchbhjx7q747jghal3l4g"
            >
              bafy2b ... al3l4g
            </a>
          </td>
          <td>
            send
          </td>
          <td>
            1020
          </td>
          <td>
            2022-08-02 09:50:30
          </td>
          <td>
            <div
              class="c3"
            >
              <div
                class="c4"
                color="var(--purple-medium)"
              >
                <span>
                  t1z225 ... 6z6wgi
                </span>
              </div>
            </div>
          </td>
          <td>
            <div
              class="c5"
              color="yellow"
            >
              <span>
                OUT
              </span>
            </div>
          </td>
          <td>
            <div
              class="c3"
            >
              <div
                class="c4"
                color="var(--purple-medium)"
              >
                <a
                  href="https://explorer-calibration.glif.link/actor/?address=t1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a"
                  rel="noreferrer noopener"
                  target="_blank"
                >
                  t1nq5k ... btuo3a
                </a>
              </div>
            </div>
          </td>
          <td>
            0.000000000000001 FIL
          </td>
        </tr>
        <tr>
          <td>
            <a
              href="/message?cid=bafy2bzaced3ub5g4v35tj7n74zsog3dmcum4tk4qmchbhjx7q747jghal3l4g"
            >
              bafy2b ... al3l4g
            </a>
          </td>
          <td>
            send
          </td>
          <td>
            1020
          </td>
          <td>
            2022-08-02 09:50:30
          </td>
          <td>
            <div
              class="c3"
            >
              <div
                class="c4"
                color="var(--purple-medium)"
              >
                <span>
                  t1z225 ... 6z6wgi
                </span>
              </div>
            </div>
          </td>
          <td>
            <div
              class="c5"
              color="yellow"
            >
              <span>
                OUT
              </span>
            </div>
          </td>
          <td>
            <div
              class="c3"
            >
              <div
                class="c4"
                color="var(--purple-medium)"
              >
                <a
                  href="https://explorer-calibration.glif.link/actor/?address=t1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a"
                  rel="noreferrer noopener"
                  target="_blank"
                >
                  t1nq5k ... btuo3a
                </a>
              </div>
            </div>
          </td>
          <td>
            0.000000000000001 FIL
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Message history it renders no message history correctly with historical data missing warning 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  gap: var(--space-l);
  margin: 1em 0;
  font-size: var(--font-size-xl);
}

.c0 .side-bar {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: var(--space-m);
  font-size: 1rem;
}

.c1 {
  margin: 0;
  color: var(--purple-medium);
}

.c2 {
  color: var(--yellow-dark);
}

<div>
  <div>
    <div
      class="c0"
    >
      <h2
        class="c1"
      >
        Transaction History
      </h2>
      <div
        class="side-bar"
      >
        <svg
          fill="none"
          height="20px"
          viewBox="0 0 22 20"
          width="22px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11 7.74982V9.99986M11 14.4999H11.0112M3.24282 19H18.7572C20.4815 19 21.5586 17.1246 20.6965 15.6249L12.9393 2.12474C12.0771 0.625088 9.92286 0.625088 9.0607 2.12474L1.30352 15.6249C0.441363 17.1246 1.5185 19 3.24282 19Z"
            stroke="var(--yellow-dark)"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <span
          class="c2"
        >
          Syncing data from Mainnet
        </span>
      </div>
    </div>
    <table>
      <caption>
        No records found 
        for Transaction History
      </caption>
      <thead>
        <tr>
          <th>
            CID
          </th>
          <th>
            Method
          </th>
          <th>
            Height
          </th>
          <th>
            Age
          </th>
          <th>
            From
          </th>
          <th />
          <th>
            To
          </th>
          <th>
            Value
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
  </div>
</div>
`;
